cleverhans/future/torch/attacks/carlini_wagner_l2.py

```diff
@@ -0,0 +1,209 @@
+  # https://github.com/pytorch/pytorch/issues/10324).
+  # This particular implementation here is not numerically
+  # stable and should be substituted w/ PyTorch's native
+  # implementation when it comes out in the future
+  arctanh = lambda x: 0.5 * torch.log((1 + x) / (1 - x))
+  x = (x - clip_min) / (clip_max - clip_min)
+  x = torch.clamp(x, 0, 1)
+  x = x * 2 - 1
+  x = arctanh(x * .999999)
+
+  # Prepare some variables
+  modifier = torch.zeros_like(x, requires_grad=True)
+  y_onehot = torch.nn.functional.one_hot(y, n_classes).to(torch.float)
+
+  # Define loss functions and optimizer
+  f_fn = lambda real, other, targeted: torch.max(
+    ((other - real) if targeted else (real - other)) + confidence,
+    torch.tensor(0.)
+  )
+  l2dist_fn = lambda x, y: torch.pow(x - y, 2).sum(list(range(len(x.size())))[1:])
+  optimizer = torch.optim.Adam([modifier], lr=lr)
+
+  # Outer loop performing binary search on const
+  for outer_step in range(binary_search_steps):
+    # Initialize some values needed for the inner loop
+    bestl2 = [INF] * len(x)
+    bestscore = [-1.] * len(x)
+
+    # Inner loop performing attack iterations
+    for i in range(max_iterations):
+      # One attack step
+      new_x = (torch.tanh(modifier + x) + 1) / 2
+      new_x = new_x * (clip_max - clip_min) + clip_min
+      logits = model_fn(new_x)
+
+      real = torch.sum(y_onehot * logits, 1)
+      other, _ = torch.max((1 - y_onehot) * logits - y_onehot * 1e4, 1)
+
+      optimizer.zero_grad()
+      f = f_fn(real, other, targeted)
+      l2 = l2dist_fn(new_x, ox)
+      loss = (const * f + l2).sum()
+      loss.backward()
+      optimizer.step()
+
+      # Update best results
+      for n, (l2_n, logits_n, new_x_n) in enumerate(zip(l2, logits, new_x)):
+        y_n = y[n]
+        succeeded = compare(logits_n, y_n, is_logits=True)
+        if l2_n < o_bestl2[n] and succeeded:
+          pred_n = torch.argmax(logits_n)
+          o_bestl2[n] = l2_n
+          o_bestscore[n] = pred_n
+          o_bestattack[n] = new_x_n
+          # l2_n < o_bestl2[n] implies l2_n < bestl2[n] so we modify inner loop variables too
+          bestl2[n] = l2_n
+          bestscore[n] = pred_n
+        elif l2_n < bestl2[n] and succeeded:
+          bestl2[n] = l2_n
+          bestscore[n] = torch.argmax(logits_n)
+
+    # Binary search step
+    for n in range(len(x)):
+      y_n = y[n]
+
+      if compare(bestscore[n], y_n) and bestscore[n] != -1:
+        # Success, divide const by two
+        upper_bound[n] = min(upper_bound[n], const[n])
+        if upper_bound[n] < 1e9:
+          const[n] = (lower_bound[n] + upper_bound[n]) / 2
+      else:
+        # Failure, either multiply by 10 if no solution found yet
+        # or do binary search with the known upper bound
+        lower_bound[n] = max(lower_bound[n], const[n])
+        if upper_bound[n] < 1e9:
+          const[n] = (lower_bound[n] + upper_bound[n]) / 2
+        else:
+          const[n] *= 10
+
+  return o_bestattack
```

Comment:
The ```requires_grad``` flag is not so much of an issue since it does not really affect anything but I do agree that the returned tensor carrying with it its entire backward graph might cause memory leaks. Fixed with d6417be. Thanks.